pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,1459 @@
+	// rancher instances containing ext tokens without idle information,
+	// without crashing. I.e. insert a default when the information is not
+	// present, instead of returning an error.
+	//
+	// idle, err := strconv.ParseInt(string(secret.Data[FieldIdleTimeout]), 10, 64)
+	// if err != nil {
+	// 	return token, err
+	// }
+	// token.Status.IdleTimeout = idle
+
+	tokenHash := string(secret.Data[FieldHash])
+	if tokenHash == "" {
+		return token, fmt.Errorf("token hash missing")
+	}
+	token.Status.TokenHash = tokenHash
+
+	authProvider := string(secret.Data[FieldAuthProvider])
+	if authProvider == "" {
+		return token, fmt.Errorf("auth provider missing")
+	}
+	token.Status.AuthProvider = authProvider
+
+	lastUpdateTime := string(secret.Data[FieldLastUpdateTime])
+	if lastUpdateTime == "" {
+		return token, fmt.Errorf("last update time missing")
+	}
+	token.Status.LastUpdateTime = lastUpdateTime
+
+	principalID := string(secret.Data[FieldPrincipalID])
+	if principalID == "" {
+		return token, fmt.Errorf("principal id missing")
+	}
+	token.Spec.PrincipalID = principalID
+
+	kubeUID := string(secret.Data[FieldUID])
+	if kubeUID == "" {
+		return token, fmt.Errorf("kube uid missing")
+	}
+	token.ObjectMeta.UID = types.UID(kubeUID)
+
+	var lastUsedAt *metav1.Time
+	lastUsedAsString := string(secret.Data[FieldLastUsedAt])
+	if lastUsedAsString != "" {
+		lastUsed, err := time.Parse(time.RFC3339, lastUsedAsString)
+		if err != nil {
+			return token, fmt.Errorf("failed to parse lastUsed data: %w", err)
+		}
+		lastUsedTime := metav1.NewTime(lastUsed)
+		lastUsedAt = &lastUsedTime
+	} // else: empty => lastUsedAt == nil
+	token.Status.LastUsedAt = lastUsedAt
+
+	if err := setExpired(token); err != nil {
+		return token, fmt.Errorf("failed to set expiration information: %w", err)
+	}
+
+	// annotations, labels - decode the respective fields and place them into the token
+
+	if err := json.Unmarshal(secret.Data[FieldLabels], &token.Labels); err != nil {
+		return token, err
+	}
+
+	if err := json.Unmarshal(secret.Data[FieldAnnotations], &token.Annotations); err != nil {
+		return token, err
+	}
+
+	return token, nil
+}
+
+// setExpired computes the expiration data (isExpired, expiresAt) from token
+// creation time and time to live and places the results into the associated
+// token fields.
+func setExpired(token *ext.Token) error {
+	if token.Spec.TTL < 0 {
+		token.Status.Expired = false
+		token.Status.ExpiresAt = ""
+		return nil
+	}
+
+	expiresAt := token.ObjectMeta.CreationTimestamp.Add(time.Duration(token.Spec.TTL) * time.Millisecond)
+	isExpired := time.Now().After(expiresAt)
+
+	eAt, err := metav1.NewTime(expiresAt).MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	// note: The marshalling puts quotes around the string. strip them
+	// before handing this to the token and yaml adding another layer
+	// of quotes around such a string
+	token.Status.ExpiresAt = string(eAt[1 : len(eAt)-1])
```

Comment:
Why are we doing this? Can `expiresAt.Format()` be used instead?

Reply:
not sure what is thinking going through json .. especially given that there is an example of `<time>.Format()` elsewhere in the store.
fixed.